### PR TITLE
sampling weight & tracestate propagation

### LINKED ISSUE #1293: Adding sampling weight to transactions and spans
Implementing https://github.com/elastic/apm/issues/307

----
Summary of changes in the spec:

On transactions & spans, add sample_rate field
`sample_rate=0`  for non sampled items (transactions only)
`sample_rate=0.1`  for sampled items (spans & transactions)

non-sampled transactions have 
- no context 
- no span
- `sampled=false`
- `sample_rate=0` 

Propagation:

add `s` attribute to the existing tracestate header (with proper rounding).

If `tracestate` is available with a valid `s` attribute (non-root transaction), then:
- sample the transaction
- propagate `sample_rate` to the related transaction  and spans

If `tracestate` is available but does not have a valid `s` attribute (non-root transaction), then:
- sample the transaction
- do not set `sample_rate` <-- this is equivalent to current implementation

If `tracestate` is not available (root transaction)
- apply configured sample rate to decide to sample or not sample
- if transaction is sampled, propagate `sample_rate` on the transaction, spans and send value through `tracestate` header if applicable

### LINKED ISSUE #1385: Implement sample_rate precision spec
You can view the spec changes [here](https://github.com/elastic/apm/pull/335).

### PR DESCRIPTION
**Pull request by SylvainJuge**
## What does this PR do?

Fixes #1293 
Fixes #1385

Implementation notes / questions
- not sure if `Sampler#getSampleRate()` is relevant as there might be non-probabilistic samplers which can't provide any meaningful value. For root transactions, we could simply get the value from configuration instead.
- sample rate rounding is applied to 1) configuration and 2) when parsing `tracestate` header
- as an optimization, we might avoid `tracestate` parsing and transmit as-is if it does not contains `es=` string, 

## Checklist

- [x] This is an enhancement of existing features, or a new feature in existing plugins
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [x] I have added tests that prove my fix is effective or that my feature works
  - [x] ~~Added an API method or config option? Document in which version this will be introduced~~
  - [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by eyalkoren — COMMENTED** [review]
Regarding your comment- "as an optimization, we might avoid tracestate parsing and transmit as-is if it does not contains es= string", I think it is more than a nice-to-have, we need to optimize already now. 

Let's try to break it:
- starting a child transaction
  - if parent passed a `tracestate`
    - if the `tracestate` header contains `es=` - only parse from its start position to the next `,`. No need to split and iterate over all of it.
    - else (`tracestate` header does not contain `es=`) - we are required to split for vendor entries because the spec does not allow more than 32 entries. However, in most cases there will be less than 32, so it should be enough to append our own to the end of it.
  - else (parent did not pass a `tracestate` header) - create a single-entry `tracestate` with our value.
- starting a root transaction - create a single-entry `tracestate` with our value. 

This is already a big optimisation, as it means we will only have to split the header if our agent is intermixed with other vendors, which is probably not the common case.

Furthermore, whenever we **add**  our `tracestate` entry, since the sample rate comes from the configuration, we can easily create a readymade `tracestate` entry and register a listener on the `transaction_sample_rate` config option to replace it when it is changed.

**Comment by SylvainJuge** [comment]
We now have
- taking the header has-is if we just need to read the value
- write and append the header to existing values
- for a given `TraceState` instance, we cache the last computed header value, thus if sample rate does not change, we don't allocate to compute header.

File: apm-agent-core/src/test/java/co/elastic/apm/agent/impl/transaction/TraceContextTest.java [hunk 1]
```diff
@@ -473,11 +481,109 @@ void testRandomValue() {
...
+    @CsvSource(delimiter = '|', value = {
+        // invalid tracestate values: just assume no sample rate is provided
+        "|null",
+        "es=|null",
+        "es=s|null",
+        "es=s:|null",
+        "es=s:a|null",
+        // valid tracestate values with sample rate
+        "es=s:1|1",
+        "es=s:0.42|0.42",
```
Comment (eyalkoren) [inline apm-agent-core/src/test/java/co/elastic/apm/agent/impl/transaction/TraceContextTest.java:557]:
```suggestion
        "es=s:0.42|0.42",
        "es=d:0.42|(maybe 1.0?)",
```

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceState.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                if (valueEnd < headerValue.length()) {
+                    char c = headerValue.charAt(valueEnd);
+                    while (valueEnd < headerValue.length() && c != VENDOR_SEPARATOR && c != ENTRY_SEPARATOR) {
+                        c = headerValue.charAt(valueEnd++);
+                    }
+                    if (valueEnd < headerValue.length()) {
+                        // end due to separator char that needs to be trimmed
+                        valueEnd--;
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (SylvainJuge) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceState.java:87]: This case is covered [by the spec](https://github.com/elastic/apm/blob/master/specs/agents/tracing-sampling.md#propagation) and we should not have a sample rate set

> For non-root transactions the agent must parse incoming tracestate headers to identify the es entry and extract the s attribute. The s attribute value should be used to populate the sample_rate field of transactions and spans. If there is no tracestate or no valid es entry with an s attribute, then the agent must omit sample_rate from non-root transactions and their spans.

However I agree with you that we probably miss a corner test case on this one.

[the remark this answers is not in the document]
Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceState.java:87]: And we should forward downstream as is, without adding it?

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceState.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+     */
+    private double sampleRate;
+
+    // temp buffer used for rewriting
+    private final StringBuilder tempBuffer = new StringBuilder();
+
+    // cache to avoid rewriting same header many times when rate does not change
+    private double lastWrittenRate = Double.NaN;
+    @Nullable
+    private String lastWrittenHeader = null;
```
[the remark this answers is not in the document]
Comment (SylvainJuge) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceState.java]: fixed in the last commit, we now have a static value stored in the sampler once for all.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceState.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        }
+        tracestate.add(writeHeader(rate));
+        sampleRate = rate;
+    }
+
+    /**
+     * @return sample rate set in tracestate header, {@literal null} if not set
+     */
+    @Nullable
+    public Double getSampleRate() {
```
[the remark this answers is not in the document]
Comment (SylvainJuge) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceState.java]: fixed in last commit, we now use `Double.NaN` as a way to express the lack of value.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceState.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (SylvainJuge) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceState.java:82]: That's how I understand the spec but that might not be consistent:
- we currently fix rounding from upstream header if required
- values < 0.0 and > 1.0, invalid or missing are just ignored, but we propagate them as-is when we might be able to "fix" them by using value from sampler.

File: apm-agent-core/src/test/java/co/elastic/apm/agent/impl/transaction/TraceContextTest.java (hunk 1 above)
Comment (SylvainJuge) [reply apm-agent-core/src/test/java/co/elastic/apm/agent/impl/transaction/TraceContextTest.java:558]: I looked quickly at ruby (ref impl) and go agents, and both consider invalid values implies same behavior as if they were simply absent.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceState.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (SylvainJuge) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceState.java:82]: I remember that the logic to add it or not depends on if the transaction is a root transaction or not:
1. if it's a root transaction: we add the sample_rate
1. if it's a child transaction and there is no valid upstream sample_rate, we do not add it so apm-server will use 1 as total distributed transaction weight (this behavior is the one used by agents that don't implement sample_rate, thus we keep it as-is for compatibility).
1. if it's a child transaction and there is a valid sample rate, we use and propagate it

Thus, to me the case where we have an invalid value (or no value) for sample rate is included in 2), but propagate as-is even if we know it's bogus is an open question.

File: apm-agent-core/src/test/java/co/elastic/apm/agent/impl/transaction/TraceContextTest.java (hunk 1 above)
Comment (eyalkoren) [reply apm-agent-core/src/test/java/co/elastic/apm/agent/impl/transaction/TraceContextTest.java:558]: OK, I guess other tests cover that

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/sampling/ProbabilitySampler.java [hunk 5]
```diff
@@ -54,10 +55,14 @@
...
     private final double sampleRate;
 
+    // Because header value only contains sampling rate, we can cache it here
+    private final String traceStateHeader;
+
     private ProbabilitySampler(double samplingRate) {
         higherBound = (long) (Long.MAX_VALUE * samplingRate);
         lowerBound = -higherBound;
         this.sampleRate = samplingRate;
+        traceStateHeader = TraceState.buildHeaderString(samplingRate);
```
Comment (eyalkoren) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/impl/sampling/ProbabilitySampler.java:65]: I don't think we need this dependency in `TraceState` - enough to be able to get `sampleRateAsString` here and use the the `StringBuilder` in `TraceState` to append to the header when required. This means both no allocation and no dependency.
Sorry for nagging, but conceptually - a sampler shouldn't be aware of `tracestate`

**Review by eyalkoren — APPROVED** [review]
Let's not keep this hanging just because of https://github.com/elastic/apm-agent-java/pull/1384#discussion_r498079421 - I'd like to avoid this dependency, but if it complicates stuff, don't bother. 
I think if `TextTracestateAppender#join()` would accept `List<CharSequence>` instead of `List<String>` that would make things simpler as you can combine `String`s and `StringBuilder`s in the tracestate list.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceState.java [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+    private static class TextTracestateAppender {
+
+        private static final TextTracestateAppender INSTANCE = new TextTracestateAppender();
+        private final ThreadLocal<StringBuilder> tracestateBuffer = new ThreadLocal<StringBuilder>();
+
+        private TextTracestateAppender() {
+        }
+
+        @Nullable
+        public String join(List<? extends CharSequence> tracestate, int tracestateSizeLimit) {
```
Comment (eyalkoren) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceState.java:192]: [minor] The optimization we lost in `TraceContext` (just forwarding a header as it was received):
```suggestion
        public String join(List<? extends CharSequence> tracestate, int tracestateSizeLimit) {
          if (tracestate.size() == 1 && tracestate.get(0) instanceof String) {
            return (String) tracestate.get(0);
          }
```

**Review by eyalkoren — APPROVED** [review]
👍  Thanks for your patience 😊

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceState.java (hunk 6 above)
Comment (SylvainJuge) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceState.java:192]: Thanks for catching this, there was one corner case to apply here though when tracestate is transmitted as-is and larger that the size limit we have to truncate it :-)

**Pull request merged**